compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5392,7 +5396,13 @@ private BSymbol getLangLibMethod(BLangInvocation iExpr, BType bType) {
 
     private void checkInvocationParamAndReturnType(BLangInvocation iExpr) {
         BType actualType = checkInvocationParam(iExpr);
-        resultType = types.checkType(iExpr, actualType, this.expType);
+        // iExpr contain NEVER_ALLOWED flag if it's in a trap expression
+        if (iExpr.flagSet != null && iExpr.flagSet.contains(Flag.NEVER_ALLOWED) &&
```

Comment:
IMO, the better approach would be to type check the invocation expression as a normal expression and then do the never validation in the relevant parent expressions/statements. If we can sort out the foreach/from usage we can get rid of the flag too.